compiler/frontend/src/org/jetbrains/kotlin/resolve/resolveUtil.kt

```diff
@@ -13,7 +13,7 @@ fun PropertyDescriptor.hasBackingField(bindingContext: BindingContext?): Boolean
     kind == CallableMemberDescriptor.Kind.FAKE_OVERRIDE -> overriddenDescriptors.any { it.hasBackingField(bindingContext) }
     source is KotlinSourceElement && bindingContext != null -> bindingContext.get(BindingContext.BACKING_FIELD_REQUIRED, this) ?: false
     compileTimeInitializer != null -> true
-//    backingField != null -> true // TODO: in case of DeserializedPropetyDescriptor this function returns incorrect result for FO field
+    backingField != null -> true
```

Comment:
We can't know if the deserialized property has a backing field or not (by design), so we always create `backingField` for deserialized properties (see the last lines in `MemberDeserializer.loadProperty`), to be able to load annotations from those fields in the frontend if needed. This change makes `hasBackingField` return true on all deserialized properties, which in turn makes every `IrLazyProperty` have a backing field. In theory this could break some assumptions in the codegen (such as that interfaces can't have fields), but also somewhat worsen the debugging experience, etc.

I propose to only add a backing field if the property is annotated with `@JvmField` instead. Actually I've just done that yesterday in my branch in https://github.com/JetBrains/kotlin/commit/b09452477c14ff173876cc91aab9451d5c9a5b81 (with a preliminary refactoring in https://github.com/JetBrains/kotlin/commit/3be5c81c393a98955e4c08b27b80888d5e330a5f). Would you mind cherry-picking these two commits instead of this change in `resolveUtil.kt`?

Reply:
These commits seem to change some origins in IrText tests, which looks suspicious to me - is there a call where the new StubGeneratorExtensions are not passed for some reason? Perhaps it's best to merge that refactoring separately then?

Reply:
Right, AbstractIrGeneratorTestCase calls Psi2IrTranslator.generateModule which was not modified to take an additional StubGeneratorExtensions parameter.

Reply:
You're right, indeed they do. I'll investigate and after fixing, will push that refactoring first, and then get back to reviewing your changes. Thanks.

Reply:
[KOTLIN-CR-3537](https://upsource.jetbrains.com/kotlin/review/KOTLIN-CR-3537)

Reply:
I've pushed my changes to master (https://github.com/JetBrains/kotlin/commit/608449f41d8725190960b80a27afd54dc632c9ad, https://github.com/JetBrains/kotlin/commit/a485a5ffd6607e6a59fef2e09180508a0bf84bea).